### PR TITLE
Add vectors

### PR DESCRIPTION
**Pull request by Grubre**
Removed useless consts for return values, added operator overloads and unit tests for vectors.

### CODE REVIEW

File: test/vector_unittest.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+TYPED_TEST(VectorTest, testing_vec2_equals)
+{
+    deen::Vector2<TypeParam> x(this->a1, this->a2);
+    deen::Vector2<TypeParam> y(this->a3, this->a4);
+    deen::Vector2<TypeParam> z(this->a1, this->a2);
+    EXPECT_TRUE(x==z);
+    EXPECT_FALSE(x==y);
+}
```
Comment (MrKubul) [inline test/vector_unittest.cpp:298]: ok

**Review by MrKubul — COMMENTED** [review]
change these details and merge afterwards

File: test/vector_unittest.cpp (hunk 1 above)
Comment (MrKubul) [reply test/vector_unittest.cpp:298]: NOT OK YOU DIDNT TEST DOT PRODUCT AND STUFF!! write unit tests for them

File: src/math/vector.hpp [hunk 2]
```diff
@@ -2,76 +2,295 @@
...
-    struct Base_vec{
-        inline const virtual T mag_sq() const = 0;
-        inline const T mag() const { return std::sqrt(mag_sq()); };
+    struct BaseVector{
+        inline virtual T magnitude_sqr() const = 0;
+        inline T magnitude() const { return std::sqrt(magnitude_sqr()); };
     };
 
 
     template<arithmetic T = double>
```
[the remark this answers is not in the document]
Comment (Grubre) [reply src/math/vector.hpp:20]: Removing this would result in unnecessary code duplication

**Comment by Grubre** [comment]
Maybe function normalized() should always return a vector of type double? For non floating point types (e.g. int)
the magnitude of the returned vector won't always be equal to 1.

**Comment by Grubre** [comment]
Same goes for magnitude() and magnitude_sqr(), the result won't always be precise (or what the user expects).

**Pull request merged**